### ISSUE TITLE
to_json mishandles strings with both \ and unprintable characters

### ISSUE THREAD

**Issue by jsiwek** [zeek/zeek#150]
Moved from https://bro-tracker.atlassian.net/browse/BIT-1959

Created by Zack Weinberg at 2018-07-25T12:46:57.194-0500:
> The `to_json` function (implemented in `base/utils/json.bro`) escapes unprintable characters using the `clean` primitive, and then manually escapes `\` and `"` with regexes.  This is wrong on two levels: the transformation performed by `clean` is irreversible, and `\x` escapes are not part of the JSON standard.  For instance, consider
> 
> ```
> @load base/utils/json
> print to_json("\"\\\x81");
> ```
> 
> Running this test program with `bro -b -C test.bro` will produce the output `"\"\\\\x81"`.  Because of the irreversible `clean` transformation, this output could correspond to either the original three-byte string (hexdumped, 22 5C 81) or a six-byte string containing two backslashes and the literal string x81 (hexdumped,  22 5C 5C 78 38 31.)  Because `\x` escapes are not part of the JSON standard, it is not enough to replace `clean` and the inner `gsub` with `escape_string`; that would produce `"\\\x81"`, which is unambiguous, but also unparseable. 
> 
> The ideal output would be `"\"\\\u0081"`.  I'm not sure how to accomplish this, considering that `gsub` does not appear to implement any way of referring to capture groups from the replacement string.  For now I'm going to change `json.bro` to read
> 
> ```
> 		case "string":
> 		return cat("\"", gsub(escape_string(v), /\"/, "\\\""), "\"");
> ```
> 
> and postprocess the JSON with a more powerful regex engine before trying to parse it.
> 
> (There is also the headache of dealing with strings with U+0000 in them, but I think it would be fair to declare that Not Your Problem.)

Comment by Zack Weinberg at 2018-07-25T12:55:06.083-0500:
> For the record, the construct that doesn't seem to be possible with `gsub`, but is possible with a "more powerful regex engine", is
> 
> ```
> perl -pe 's/(?<!\\)((?:\\\\)*\\)x/${1}u00/g'
> ```

Comment by Robin Sommer at 2018-08-01T12:45:08.143-0500:
> to_json() should probably become a bif so that we can write it in C++ and "do the right thing". This is part of a broader question of potentially making Bro more JSON-friendly. Having well-defined JSON de-/serialization of Bro values would be nice.
> 
> And yeah, our regexps don't support capture groups. That's another item on the roadmap: consider replacing Bro's engine with one of the modern libraries; that needs some thought & research.

**Comment by rsmmr** [comment]
Not quite sure which piece here triggered the 3.0 tag, but I'd suggest we tackle the part on moving `to_json()` into C++ bif.

**Comment by jsiwek** [comment]
> Not quite sure which piece here triggered the 3.0 tag

Think it was mostly added in the context of looking for any "bad" bugs and this one seemed like it's a fairly broken usability issue.  But I wouldn't consider it a showstopper for 3.0.

**Comment by timwoj** [comment]
At a quick glance, this is basically wanting to replace the `to_json` method in `scripts/base/utils/json.zeek` with one in a bif that's written in C++. Is that correct?

**Comment by jsiwek** [comment]
@timwoj yeah, my understanding of the scope here:  re-implement `to_json` as BIF/C++ and also properly escape characters in strings that require it

**Comment by timwoj** [comment]
A couple questions so far:

- How should a `set` be represented? They're technically a table internally, but it's really just a list of values similar to a vector.
- Is it possible to call other non-C++ bif methods from inside C++ in a bif? For example, I could really use the `record_fields` method.

**Comment by rsmmr** [comment]
- Representing a set like a vector sounds good to me
- Yes, you can call them, though it's not pretty: look at `build/src/bro.bif.func_h` to see how `record_fields` ends up looking at the C++ level. In this specific case it might be easier to just replicate the code or factor it out into a separate C++ function.

**Comment by timwoj** [comment]
How should a table be represented when the key isn't an elementary type? Right now it looks like `json.zeek` just converts the key object into json and uses a string of that as the key for the table entry, but that seems ugly at best.

**Comment by rsmmr** [comment]
For atomic keys, that sounds right: just convert to string. For composite keys, not sure ... converting them to string isn't great, but I don't see much better either; other than just declining to work on such tables.

**Comment by jsiwek** [comment]
For either an atomic key or a composite key, a Zeek `table` can be defined as a JSON `object` whose keys are a JSON `string` containing the content of any arbitrary Zeek value after its conversion to a JSON `value`.  Doesn't seem that "ugly" to me -- it's recursively consistent (plus can't think of another way to do it).

Not sure if there's other gotchas, but inner JSON `string` values within the key at least need to escape their use of double-quote characters.

**Comment by timwoj** [comment]
```
local g = table(
    [192.168.0.2, 22/tcp] = "ssh",
    [192.168.0.3, 80/tcp] = "http"
);
```

turns into

`{"[\"192.168.0.2\",\"22/tcp\"]":"ssh","[\"192.168.0.3\",\"80/tcp\"]":"http"}`

That looks ok and is readable enough.

**Comment by timwoj** [comment]
Is there a reason why the field replacement doesn't get passed down to further calls of `to_json` in the existing code?

```
# replace the escape pattern in the field.
if( field_escape_pattern in field )
	field = cat(sub(field, field_escape_pattern, ""));
```

It only uses the user-provided pattern on the first call, and then uses the default of `/^_` for any subsequent calls.

**Comment by timwoj** [comment]
Fixed in 1f329ad541fa8c199e3453bdb5bd9ecb5d139b53
